plugins/policy-recommendation/policy_recommendation_job.py

```diff
@@ -0,0 +1,843 @@
+from urllib.parse import urlparse
+
+import antrea_crd
+from policy_recommendation_utils import *
+
+# Column names of flow record table in Clickhouse database used in recommendation job
+FLOW_TABLE_COLUMNS = [
+    'sourcePodNamespace',
+    'sourcePodLabels',
+    'destinationIP',
+    'destinationPodNamespace',
+    'destinationPodLabels',
+    'destinationServicePortName',
+    'destinationTransportPort',
+    'protocolIdentifier',
+]
+
+NAMESPACE_ALLOW_LIST = [
+    'kube-system',
+    'flow-aggregator',
+    'flow-visibility'
+]
+
+ROW_DELIMITER = "#"
+PEER_DELIMITER = "|"
+DEFAULT_POLICY_PRIORITY = 5
+
+MEANINGLESS_LABELS = [
+    "pod-template-hash",
+    "controller-revision-hash",
+    "pod-template-generation"
+]
+
+def get_flow_type(destinationServicePortName, destinationPodLabels):
+    if destinationServicePortName != "":
+        return "pod_to_svc"
+    elif destinationPodLabels != "":
+        return "pod_to_pod"
+    else:
+        return "pod_to_external"
+
+def remove_meaningless_labels(podLables):
+    try:
+        labels_dict = json.loads(podLables)
+    except Exception as e:
+        print("Error {}: labels {} is not in json format".format(e, podLables))
+        return ""
+    for key in list(labels_dict):
+        if key in MEANINGLESS_LABELS:
+            labels_dict.pop(key)
+    return json.dumps(labels_dict, sort_keys=True)
+
+def get_protocol_string(protocolIdentifier):
+    if protocolIdentifier == 6:
+        return "TCP"
+    elif protocolIdentifier == 17:
+        return "UDP"
+    else:
+        return "UNKNOWN"
+
+def map_flow_to_egress(flow, k8s=False):
+    src = ROW_DELIMITER.join([flow.sourcePodNamespace, flow.sourcePodLabels])
+    if flow.flowType == "pod_to_external":
+        dst = ROW_DELIMITER.join([flow.destinationIP, str(flow.destinationTransportPort), get_protocol_string(flow.protocolIdentifier)])
+    elif flow.flowType == "pod_to_svc" and not k8s:
+        # K8s policies don't support Pod to Service rules
+        svc_ns, svc_name = flow.destinationServicePortName.partition(':')[0].split('/')
+        dst = ROW_DELIMITER.join([svc_ns, svc_name])
+    else:
+        dst = ROW_DELIMITER.join([flow.destinationPodNamespace, flow.destinationPodLabels, str(flow.destinationTransportPort), get_protocol_string(flow.protocolIdentifier)])
+    return (src, ("", dst))
+
+def map_flow_to_egress_svc(flow):
+    src = ROW_DELIMITER.join([flow.sourcePodNamespace, flow.sourcePodLabels])
+    dst = ROW_DELIMITER.join([flow.destinationServicePortName, str(flow.destinationTransportPort), get_protocol_string(flow.protocolIdentifier)])
+    return (src, dst)
+
+def map_flow_to_ingress(flow):
+    src = ROW_DELIMITER.join([flow.sourcePodNamespace, flow.sourcePodLabels, str(flow.destinationTransportPort), get_protocol_string(flow.protocolIdentifier)])
+    dst = ROW_DELIMITER.join([flow.destinationPodNamespace, flow.destinationPodLabels])
+    return (dst, (src, ""))
+
+def combine_network_peers(a, b):
```

Comment:
This routine might deserve some comment...

Reply:
Added.